### PR TITLE
[StarRocks On ES] Support array format.

### PR DESCRIPTION
**Pull request by Smith-Cruise**
Support array format in ES.

## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by Smith-Cruise** [comment]
run starrocks_clang-format

File: be/src/exec/es/es_scroll_parser.cpp [hunk 1]
```diff
@@ -81,6 +82,14 @@ std::string json_value_to_string(const rapidjson::Value& value) {
         }                                                                     \
     } while (false)
 
+#define RETURN_ERROR_IF_COL_IS_INVALID_ARRAY_FORMAT(col)             \
+    do {                                                             \
+        std::stringstream ss;                                        \
+        ss << "Invalid array format"                                 \
+           << "; Document slice is : " << json_value_to_string(col); \
+        return Status::RuntimeError(ss.str());                       \
```
Comment (imay) [inline be/src/exec/es/es_scroll_parser.cpp:90]: why not use fmt::format() directly?

File: be/src/exec/es/es_scroll_parser.cpp [hunk 2]
```diff
@@ -504,6 +517,43 @@ Status ScrollParser::_append_bool_val(const rapidjson::Value& col, Column* colum
 
     return Status::OK();
 }
+
+Status ScrollParser::_append_array_val(const rapidjson::Value& col, const TypeDescriptor& type_desc, Column* column,
```
[the remark this answers is not in the document]
Comment (Smith-Cruise) [reply be/src/exec/es/es_scroll_parser.cpp:521]: I also think `col` is hard to understand, but all existing methods use `col`.

[the remark this answers is not in the document]
Comment (imay) [reply be/src/exec/es/es_scroll_parser.cpp:521]: You can refactor them in another PR.

File: be/src/exec/es/es_scroll_parser.cpp [hunk 3]
```diff
@@ -504,6 +509,44 @@ Status ScrollParser::_append_bool_val(const rapidjson::Value& col, Column* colum
...
+                                       bool pure_doc_value) {
+    // Array type must have child type.
+    const auto& child_type = type_desc.children[0];
+    DCHECK(child_type.type != INVALID_TYPE);
+
+    // In ElasticSearch, n-dimensional array will be flattened into one-dimensional array.
+    // https://www.elastic.co/guide/en/elasticsearch/reference/8.3/array.html
+    // TODO: We should do schema validation in FE.
+    if (!col.IsArray() || child_type.type == TYPE_ARRAY) {
```
Comment (wuyunfeng) [inline be/src/exec/es/es_scroll_parser.cpp:522]: let us discuss this online, this process is not right

**Review by wuyunfeng — CHANGES_REQUESTED** [review]

File: be/src/exec/es/es_scroll_parser.cpp (hunk 3 above)
Comment (RowenWoo) [inline be/src/exec/es/es_scroll_parser.cpp:522]: let us schedule a online meeting for this feature, this process is not right

**Review by RowenWoo — CHANGES_REQUESTED** [review]

File: be/src/exec/es/es_scroll_parser.cpp [hunk 4]
```diff
@@ -504,6 +514,75 @@ Status ScrollParser::_append_bool_val(const rapidjson::Value& col, Column* colum
...
+        auto* nullable_column = down_cast<NullableColumn*>(column);
+        auto* data_column = nullable_column->data_column().get();
+        NullData& null_data = nullable_column->null_column_data();
+        null_data.push_back(0);
+        array = down_cast<ArrayColumn*>(data_column);
+    } else {
+        array = down_cast<ArrayColumn*>(column);
+    }
+
+    auto* offsets = array->offsets_column().get();
```
Comment (imay) [inline be/src/exec/es/es_scroll_parser.cpp:545]: Should define `offsets` when using it to understand easily.

File: be/src/exec/es/es_scroll_parser.cpp [hunk 5]
```diff
@@ -504,6 +514,75 @@ Status ScrollParser::_append_bool_val(const rapidjson::Value& col, Column* colum
...
+                                                 Column* column) {
+    for (auto& item : col.GetArray()) {
+        RETURN_IF_ERROR(_append_value_from_json_val(column, child_type_desc, item, true));
+    }
+    return Status::OK();
+}
+
+Status ScrollParser::_append_source_array_val(const rapidjson::Value& col, const TypeDescriptor& child_type_desc,
+                                              Column* column) {
+    if (col.IsNull()) {
```
Comment (imay) [inline be/src/exec/es/es_scroll_parser.cpp:569]: If col is null, is it treated as a null or an empty array?

Comment (Smith-Cruise) [reply be/src/exec/es/es_scroll_parser.cpp:569]: empty array, I have tested this situation in ut.

https://github.com/StarRocks/starrocks/blob/066d98ceaa748d183ed91ea56b89b19eb1c491ab/be/test/exec/es/es_scroll_parser_test.cpp#L148-L171

Comment (imay) [reply be/src/exec/es/es_scroll_parser.cpp:569]: I think we should return null if it doesn't exist in the ES

Comment (RowenWoo) [inline be/src/exec/es/es_scroll_parser.cpp:569]: fill_chunk() method has checked col.IsNull, maybe here is DCHECK()

Comment (Smith-Cruise) [reply be/src/exec/es/es_scroll_parser.cpp:569]: We've already discussed it @RowenWoo .
We should keep the same behavior with doc_values.

File: be/src/exec/es/es_scroll_parser.cpp [hunk 6]
```diff
@@ -504,6 +514,75 @@ Status ScrollParser::_append_bool_val(const rapidjson::Value& col, Column* colum
...
+                                       bool pure_doc_value) {
+    // Array type must have child type.
+    const auto& child_type = type_desc.children[0];
+    DCHECK(child_type.type != INVALID_TYPE);
+
+    // In ElasticSearch, n-dimensional array will be flattened into one-dimensional array.
+    // https://www.elastic.co/guide/en/elasticsearch/reference/8.3/array.html
+    // So we do not support user to create nested array column.
+    // TODO: We should prevent user to create nested array column in FE, but we don't do any schema validation now.
+    if (child_type.type == TYPE_ARRAY) {
```
Comment (RowenWoo) [inline be/src/exec/es/es_scroll_parser.cpp:528]: should remove this `if` branch?

File: be/src/exec/es/es_scroll_parser.cpp [hunk 7]
```diff
@@ -504,6 +514,75 @@ Status ScrollParser::_append_bool_val(const rapidjson::Value& col, Column* colum
...
+        std::string str = fmt::format("Invalid array format; Document slice is: {}.", json_value_to_string(col));
+        return Status::RuntimeError(str);
+    }
+
+    ArrayColumn* array = nullptr;
+
+    if (column->is_nullable()) {
```
Comment (RowenWoo) [inline be/src/exec/es/es_scroll_parser.cpp:535]: nullable is judged outside this method

**Review by RowenWoo — CHANGES_REQUESTED** [review]

File: be/src/exec/es/es_scroll_parser.cpp (hunk 5 above)
Comment (RowenWoo) [reply be/src/exec/es/es_scroll_parser.cpp:569]: if the column does not exists in `_source` or `doc_value`, we return null, i think maybe this is the only one scenario

Comment (RowenWoo) [reply be/src/exec/es/es_scroll_parser.cpp:569]: "fields": { 
     //         "array": [] 
     //     } ,

this happen after version 7.x

File: be/src/exec/es/es_scroll_parser.cpp (hunk 6 above)
Comment (Smith-Cruise) [reply be/src/exec/es/es_scroll_parser.cpp:528]: This if used for check nested array column.

File: be/src/exec/es/es_scroll_parser.h [hunk 8]
```diff
@@ -33,39 +33,59 @@ class ScrollParser {
...
+    static Status _append_float_val(const rapidjson::Value& col, Column* column, bool pure_doc_value);
 
-    Status _append_bool_val(const rapidjson::Value& col, Column* column, bool pure_doc_value);
+    static Status _append_bool_val(const rapidjson::Value& col, Column* column, bool pure_doc_value);
 
     template <PrimitiveType type, typename T = RunTimeCppType<type>>
-    Status _append_date_val(const rapidjson::Value& col, Column* column, bool pure_doc_value);
+    static Status _append_date_val(const rapidjson::Value& col, Column* column, bool pure_doc_value);
+
+    // In ES, array maybe stored as two different format, we should deal with it separately.
```
Comment (RowenWoo) [inline be/src/exec/es/es_scroll_parser.h:64]: The representation of array types in `_source` and `doc_value` (column storage) is inconsistent in Elasticsearch
, we need to do different processing to show consistent behavior

File: be/src/exec/es/es_scroll_parser.cpp [hunk 9]
```diff
@@ -504,6 +514,75 @@ Status ScrollParser::_append_bool_val(const rapidjson::Value& col, Column* colum
...
+    } else {
+        RETURN_IF_ERROR(_append_source_array_val(col, child_type, elements));
+    }
+
+    size_t new_size = elements->size();
+    offsets->append(new_size);
+    return Status::OK();
+}
+
+Status ScrollParser::_append_doc_value_array_val(const rapidjson::Value& val, const TypeDescriptor& child_type_desc,
```
Comment (RowenWoo) [inline be/src/exec/es/es_scroll_parser.cpp:559]:
```suggestion
Status ScrollParser::_append_array_val_from_docvalue(const rapidjson::Value& val, const TypeDescriptor& child_type_desc,
```

File: be/src/exec/es/es_scroll_parser.cpp [hunk 10]
```diff
@@ -504,6 +514,75 @@ Status ScrollParser::_append_bool_val(const rapidjson::Value& col, Column* colum
...
+                                                 Column* column) {
+    for (auto& item : val.GetArray()) {
+        RETURN_IF_ERROR(_append_value_from_json_val(column, child_type_desc, item, true));
+    }
+    return Status::OK();
+}
+
+Status ScrollParser::_append_source_array_val(const rapidjson::Value& val, const TypeDescriptor& child_type_desc,
```
Comment (RowenWoo) [inline be/src/exec/es/es_scroll_parser.cpp:567]:
```suggestion
Status ScrollParser::_append_array_val_from_source(const rapidjson::Value& val, const TypeDescriptor& child_type_desc,
```

**Review by RowenWoo — CHANGES_REQUESTED** [review]

**Review by imay — APPROVED** [review]

**Review by RowenWoo — APPROVED** [review]
LGTM

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Smith-Cruise** [comment]
@Mergifyio backport branch-2.1

**Comment by Smith-Cruise** [comment]
@Mergifyio backport branch-2.3

**Comment by Smith-Cruise** [comment]
@Mergifyio backport branch-2.4

**Pull request merged**